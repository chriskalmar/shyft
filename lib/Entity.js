
import {
  passOrThrow,
  // isDataType,
  // resolveFunctionMap,
} from './util';


class Entity {

  constructor (setup = {}) {

    const {
      name,
      description,
      attributes,
    } = setup

    passOrThrow(name, 'Missing entity name')
    passOrThrow(description, `Missing description for entity '${name}'`)

    this.name = name
    this.description = description
    this._attributesMap = attributes
  }


  getAttributes () {
    return this._attributes || (this._attributes = this._processAttributeMap())
  }


  _processAttributeMap () {
    // const attributeMap = resolveFunctionMap(this._attributesMap);
  }


  toString() {
    return this.name
  }

}


export default Entity
