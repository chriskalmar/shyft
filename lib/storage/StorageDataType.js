
import {
  passOrThrow,
  isFunction,
} from '../util';


class StorageDataType {

  constructor (setup = {}) {

    const {
      name,
      description,
      nativeDataType,
      isSortable,
      serialize,
      parse,
    } = setup

    passOrThrow(name, () => 'Missing storage data type name')
    passOrThrow(description, () => `Missing description for storage data type '${name}'`)
    passOrThrow(nativeDataType, () => `Missing native data type for storage data type '${name}'`)

    passOrThrow(
      isFunction(serialize),
      () => `Storage data type '${name}' has an invalid serialize function`
    )

    passOrThrow(
      !parse || isFunction(parse),
      () => `Storage data type '${name}' has an invalid parse function`
    )

    this.name = name
    this.description = description
    this.nativeDataType = nativeDataType
    this.isSortable = !!isSortable
    this.serialize = serialize
    this.parse = parse || (value => value)
  }


  toString() {
    return this.name
  }

}


export default StorageDataType


export const isStorageDataType = (obj) => {
  return (obj instanceof StorageDataType)
}
