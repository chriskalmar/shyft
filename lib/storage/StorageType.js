
import { passOrThrow } from '../util';


class StorageType {

  constructor (setup = {}) {

    const {
      name,
      description,
    } = setup

    passOrThrow(name, () => 'Missing storage type name')
    passOrThrow(description, () => `Missing description for storage type '${name}'`)

    this.name = name
    this.description = description
  }


  findOne() {
    throw new Error(`Storage type '${this.name} needs to implement findOne()`)
  }


  find() {
    throw new Error(`Storage type '${this.name} needs to implement find()`)
  }


  // filter capabilities
  // sort capabilities
  // ...


  toString() {
    return this.name
  }

}


export default StorageType
