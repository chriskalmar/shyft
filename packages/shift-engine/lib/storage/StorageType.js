
import {
  passOrThrow,
  isFunction,
} from '../util';

import { isDataType } from '../datatype/DataType';
import { isStorageDataType } from './StorageDataType';

class StorageType {

  constructor (setup = {}) {

    const {
      name,
      description,
      findOne,
      findOneByValues,
      find,
      count,
      mutate,
    } = setup

    passOrThrow(name, () => 'Missing storage type name')
    passOrThrow(description, () => `Missing description for storage type '${name}'`)

    passOrThrow(
      isFunction(findOne),
      () => `Storage type '${name}' needs to implement findOne()`
    )

    passOrThrow(
      isFunction(findOneByValues),
      () => `Storage type '${name}' needs to implement findOneByValues()`
    )

    passOrThrow(
      isFunction(find),
      () => `Storage type '${name}' needs to implement find()`
    )

    passOrThrow(
      isFunction(count),
      () => `Storage type '${name}' needs to implement count()`
    )

    passOrThrow(
      isFunction(mutate),
      () => `Storage type '${name}' needs to implement mutate()`
    )


    this.name = name
    this.description = description
    this.findOne = findOne
    this.findOneByValues = findOneByValues
    this.find = find
    this.count = count
    this.mutate = mutate

    this._dataTypeMap = {}
  }


  addDataTypeMap(schemaDataType, storageDataType) {

    passOrThrow(
      isDataType(schemaDataType),
      () => `Provided schemaDataType is not a valid data type in '${this.name}', ` +
        `got this instead: ${String(schemaDataType)}`
    )

    passOrThrow(
      isStorageDataType(storageDataType),
      () => `Provided storageDataType for '${schemaDataType}' is not a valid storage data type in '${this.name}', ` +
        `got this instead: ${String(storageDataType)}`
    )

    passOrThrow(
      !this._dataTypeMap[ schemaDataType.name ],
      () => `Data type mapping for '${schemaDataType.name}' already registered with storage type '${this.name}'`
    )

    this._dataTypeMap[ schemaDataType.name ] = storageDataType
  }


  convertToStorageDataType (schemaDataType) {

    passOrThrow(
      isDataType(schemaDataType),
      () => `Provided schemaDataType is not a valid data type in storage type '${this.name}', ` +
        `got this instead: ${String(schemaDataType)}`
    )

    passOrThrow(
      this._dataTypeMap[ schemaDataType.name ],
      () => `No data type mapping found for '${schemaDataType.name}' in storage type '${this.name}'`
    )

    return this._dataTypeMap[ schemaDataType.name ]
  }


  setStorageInstanceContextKey (storageInstanceContextKey) {
    this.storageInstanceContextKey = storageInstanceContextKey
  }


  getStorageInstanceFromContext (context) {
    passOrThrow(
      this.storageInstanceContextKey,
      () => `Storage instance context key not set for storage type '${this.name}'`
    )

    passOrThrow(
      context[ this.storageInstanceContextKey ],
      () => `Storage instance not found at context key for storage type '${this.name}'`
    )

    return context[ this.storageInstanceContextKey ]
  }


  toString() {
    return this.name
  }

}


export default StorageType


export const isStorageType = (obj) => {
  return (obj instanceof StorageType)
}
