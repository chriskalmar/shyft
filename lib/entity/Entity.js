
import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  isFunction,
} from '../util';

import {
  ATTRIBUTE_NAME_PATTERN,
  attributeNameRegex,
} from '../constants';

import { isDataType } from '../datatype/DataType';
import { isStorageType } from '../storage/StorageType';

import { StorageTypeMemory } from '../storage/StorageTypeMemory';

class Entity {

  constructor (setup = {}) {

    const {
      name,
      description,
      attributes,
      storageType,
    } = setup

    passOrThrow(name, () => 'Missing entity name')
    passOrThrow(description, () => `Missing description for entity '${name}'`)
    passOrThrow(attributes, () => `Missing attributes for entity '${name}'`)

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () => `Entity '${name}' needs an attribute definition as a map or a function returning such a map`
    )

    this.name = name
    this.description = description
    this._attributesMap = attributes

    if (storageType) {
      passOrThrow(
        isStorageType(storageType),
        () => `Entity '${name}' needs a valid storage type (defaults to 'StorageTypeMemory')`
      )
    }
    else {
      this.storageType = StorageTypeMemory
      this.isFallbackStorageType = true
      this._exposeStorageAccess()
    }

  }


  _injectStorageTypeBySchema (storageType) {

    passOrThrow(
      isStorageType(storageType),
      () => `Provided storage type to entity '${this.name}' is invalid`
    )

    if (this.isFallbackStorageType) {
      this.storageType = storageType
      this._exposeStorageAccess()
    }
  }


  _exposeStorageAccess () {
    this.findOne = this.storageType.findOne
    this.find = this.storageType.find
  }



  getAttributes () {
    return this._attributes || (this._attributes = this._processAttributeMap())
  }



  _attachSystemAttributes (attributeMap) {

    const systemAttributes = this.storageType.systemAttributes

    if (systemAttributes) {
      const attributeNames = Object.keys(systemAttributes)

      attributeNames.forEach((attributeName) => {

        passOrThrow(
          !attributeMap[ attributeName ],
          () => `Attribute name collision with system attribute '${attributeName}' in entity '${this.name}'`
        )

        attributeMap[ attributeName ] = systemAttributes[ attributeName ]
      })
    }

  }



  _processAttributeMap () {

    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () => `Attribute definition function for entity '${this.name}' does not return a map`
    )

    this._attachSystemAttributes(attributeMap)

    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Entity '${this.name}' has no attributes defined`
    )

    const resultAttributes = {}

    attributeNames.forEach((attributeName) => {

      passOrThrow(
        attributeNameRegex.test(attributeName),
        () => `Invalid attribute name '${attributeName}' in entity '${this.name}' (Regex: /${ATTRIBUTE_NAME_PATTERN}/)`
      )

      const attribute = {
        ...attributeMap[attributeName],
        required: !!attributeMap[attributeName].required,
        name: attributeName
      }

      passOrThrow(attribute.description, () => `Missing description for '${this.name}.${attributeName}'`)

      passOrThrow(
        isDataType(attribute.type) || (attribute.type instanceof Entity),
        () => `'${this.name}.${attributeName}' has invalid data type '${String(attribute.type)}'`
      )

      passOrThrow(
        !attribute.resolve || isFunction(attribute.resolve),
        () => `'${this.name}.${attributeName}' has an invalid resolve function'`
      )

      passOrThrow(
        !attribute.defaultValue || isFunction(attribute.defaultValue),
        () => `'${this.name}.${attributeName}' has an invalid defaultValue function'`
      )

      resultAttributes[ attributeName ] = attribute
    })

    return resultAttributes
  }


  toString() {
    return this.name
  }

}


export default Entity


export const isEntity = (obj) => {
  return (obj instanceof Entity)
}
