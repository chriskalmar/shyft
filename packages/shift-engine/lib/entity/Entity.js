
import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  isArray,
  isFunction,
  mapOverProperties,
} from '../util';

import {
  ATTRIBUTE_NAME_PATTERN,
  attributeNameRegex,
} from '../constants';

import { isIndex, INDEX_UNIQUE } from '../index/Index';
import Mutation, {
  isMutation,
  defaultEntityMutations,
  MUTATION_TYPE_CREATE,
} from '../mutation/Mutation';

import { isPermission } from '../permission/Permission';
import { isDataType } from '../datatype/DataType';
import { isStorageType } from '../storage/StorageType';
import { StorageTypeNull } from '../storage/StorageTypeNull';

import {
  systemAttributePrimary,
  systemAttributesTimeTracking,
  systemAttributesUserTracking,
} from './systemAttributes';



class Entity {

  constructor (setup = {}) {

    const {
      name,
      description,
      attributes,
      storageType,
      isUserEntity,
      includeTimeTracking,
      includeUserTracking,
      indexes,
      mutations,
      permissions,
    } = setup

    passOrThrow(name, () => 'Missing entity name')
    passOrThrow(description, () => `Missing description for entity '${name}'`)
    passOrThrow(attributes, () => `Missing attributes for entity '${name}'`)

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () => `Entity '${name}' needs an attribute definition as a map or a function returning such a map`
    )

    this.name = name
    this.description = description
    this.isUserEntity = !!isUserEntity
    this.includeTimeTracking = !!includeTimeTracking
    this.includeUserTracking = !!includeUserTracking
    this._attributesMap = attributes
    this._primaryAttribute = null
    this.referencedByEntities = []

    if (storageType) {
      passOrThrow(
        isStorageType(storageType),
        () => `Entity '${name}' needs a valid storage type (defaults to 'StorageTypeNull')`
      )
    }
    else {
      this.storageType = StorageTypeNull
      this.isFallbackStorageType = true
      this._exposeStorageAccess()
    }



    if (indexes) {

      this.indexes = indexes

      passOrThrow(
        isArray(indexes),
        () => `Entity '${name}' indexes definition needs to be an array of indexes`
      )

      indexes.map((index, idx) => {
        passOrThrow(
          isIndex(index),
          () => `Invalid index defintion for entity '${name}' at position '${idx}'`
        )

      })
    }


    if (mutations) {

      this.mutations = mutations

      passOrThrow(
        isArray(mutations),
        () => `Entity '${name}' mutations definition needs to be an array of mutations`
      )

      mutations.map((mutation, idx) => {
        passOrThrow(
          isMutation(mutation),
          () => `Invalid mutation defintion for entity '${name}' at position '${idx}'`
        )

      })
    }


    if (permissions) {

      this.permissions = permissions

      passOrThrow(
        isMap(permissions),
        () => `Entity '${name}' permissions definition needs to be an object`
      )


      if (permissions.read) {
        passOrThrow(
          isPermission(permissions.read),
          () => `Invalid read permission defintion for entity '${name}'`
        )
      }

      if (permissions.mutations) {
        passOrThrow(
          isMap(permissions.mutations),
          () => `Entity '${name}' permissions definition for mutations needs to be a map of mutations and permissions`
        )

        const mutationNames = Object.keys(permissions.mutations);
        mutationNames.map((mutationName, idx) => {
          passOrThrow(
            isPermission(permissions.mutations[ mutationName ]),
            () => `Invalid mutation permission defintion for entity '${name}' at position '${idx}'`
          )

        })
      }

    }
  }


  _injectStorageTypeBySchema (storageType) {

    passOrThrow(
      isStorageType(storageType),
      () => `Provided storage type to entity '${this.name}' is invalid`
    )

    if (this.isFallbackStorageType) {
      this.storageType = storageType
      this._exposeStorageAccess()
    }
  }


  _exposeStorageAccess () {
    this.findOne = this.storageType.findOne
    this.find = this.storageType.find
  }



  getAttributes () {
    if (this._attributes) {
      return this._attributes
    }

    const ret = this._attributes = this._processAttributeMap()
    this._processIndexes()
    this._processMutations()
    return ret
  }


  getMutations () {
    return this.mutations
  }


  _collectSystemAttributes (attributeMap) {

    const list = []

    if (!this.getPrimaryAttribute()) {
      this._checkSystemAttributeNameCollision(attributeMap, systemAttributePrimary.name)
      attributeMap[ systemAttributePrimary.name ] = systemAttributePrimary
      list.push(systemAttributePrimary.name)
    }

    if (this.includeTimeTracking) {
      systemAttributesTimeTracking.map(attribute => {
        this._checkSystemAttributeNameCollision(attributeMap, attribute.name)
        attributeMap[ attribute.name ] = attribute
        list.push(attribute.name)
      })
    }

    if (this.includeUserTracking && !this.isUserEntity) {
      systemAttributesUserTracking.map(attribute => {
        this._checkSystemAttributeNameCollision(attributeMap, attribute.name)
        attributeMap[ attribute.name ] = attribute
        list.push(attribute.name)
      })
    }

    return list
  }



  _checkSystemAttributeNameCollision (attributeMap, attributeName) {
    passOrThrow(
      !attributeMap[ attributeName ],
      () => `Attribute name collision with system attribute '${attributeName}' in entity '${this.name}'`
    )
  }


  _processAttribute (rawAttribute, attributeName) {

    passOrThrow(
      attributeNameRegex.test(attributeName),
      () => `Invalid attribute name '${attributeName}' in entity '${this.name}' (Regex: /${ATTRIBUTE_NAME_PATTERN}/)`
    )

    const attribute = {
      ...rawAttribute,
      isPrimary: !!rawAttribute.isPrimary,
      isUnique: !!rawAttribute.isPrimary,
      required: !!rawAttribute.required,
      hidden: !!rawAttribute.hidden,
      name: attributeName
    }

    passOrThrow(attribute.description, () => `Missing description for '${this.name}.${attributeName}'`)

    passOrThrow(
      isDataType(attribute.type) || (attribute.type instanceof Entity),
      () => `'${this.name}.${attributeName}' has invalid data type '${String(attribute.type)}'`
    )

    if (attribute.targetAttributesMap) {
      passOrThrow(
        attribute.type instanceof Entity,
        () => `'${this.name}.${attributeName}' cannot have a targetAttributesMap as it is not a reference`
      )

      passOrThrow(
        isMap(attribute.targetAttributesMap),
        () => `targetAttributesMap for '${this.name}.${attributeName}' needs to be a map`
      )

      const localAttributeNames = Object.keys(attribute.targetAttributesMap);
      localAttributeNames.map(localAttributeName => {
        const targetAttribute = attribute.targetAttributesMap[ localAttributeName ]

        passOrThrow(
          isMap(targetAttribute) && targetAttribute.name && targetAttribute.type,
          () => `targetAttributesMap for '${this.name}.${attributeName}' needs to be a map between local and target attributes`
        )

        // check if attribute is found in target entity
        attribute.type.referenceAttribute(targetAttribute.name)
      })
    }

    if (attribute.isPrimary) {
      passOrThrow(
        !this._primaryAttribute,
        () => `'${this.name}.${attributeName}' cannot be set as primary attribute,` +
          `'${this._primaryAttribute.name}' is already the primary attribute`
      )

      passOrThrow(
        isDataType(attribute.type),
        () => `Primary attribute '${this.name}.${attributeName}' has invalid data type '${String(attribute.type)}'`
      )

      this._primaryAttribute = attribute
    }

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`
    )

    passOrThrow(
      !attribute.defaultValue || isFunction(attribute.defaultValue),
      () => `'${this.name}.${attributeName}' has an invalid defaultValue function'`
    )

    return attribute
  }


  _processAttributeMap () {

    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () => `Attribute definition function for entity '${this.name}' does not return a map`
    )


    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Entity '${this.name}' has no attributes defined`
    )

    const resultAttributes = {}

    attributeNames.forEach((attributeName) => {
      resultAttributes[ attributeName ] = this._processAttribute(attributeMap[ attributeName ], attributeName)
    })

    attributeNames.forEach((attributeName) => {
      const attribute = resultAttributes[ attributeName ]

      if (attribute.targetAttributesMap) {
        const localAttributeNames = Object.keys(attribute.targetAttributesMap);
        localAttributeNames.map(localAttributeName => {
          passOrThrow(
            resultAttributes[ localAttributeName ],
            () => `Unknown local attribute '${localAttributeName}' used in targetAttributesMap ` +
              `for '${this.name}.${attributeName}'`
          )

        })
      }
    })


    const systemAttributeNames = this._collectSystemAttributes(attributeMap)

    systemAttributeNames.forEach((attributeName) => {
      resultAttributes[ attributeName ] = this._processAttribute(attributeMap[ attributeName ], attributeName)
      resultAttributes[ attributeName ].isSystemAttribute = true
    })

    return resultAttributes
  }


  getPrimaryAttribute () {
    return this._primaryAttribute
  }


  referenceAttribute (attributeName) {
    const attributes = this.getAttributes()

    passOrThrow(
      attributes[ attributeName ],
      () => `Cannot reference attribute '${this.name}.${attributeName}' as it does not exist`
    )

    return attributes[ attributeName ]
  }


  _processIndexes () {
    if (this.indexes) {

      this.indexes.map((index) => {
        index.attributes.map((attributeName) => {

          passOrThrow(
            this._attributes[ attributeName ],
            () => `Cannot use attribute '${this.name}.${attributeName}' in index as it does not exist`
          )

          if (index.type === INDEX_UNIQUE && index.attributes.length === 1) {
            this._attributes[ attributeName ].isUnique = true
          }

        })
      })
    }
  }


  _processMutations () {
    const _self = this

    const coreAttributeNames = []
    const requiredAttributeNames = []

    mapOverProperties(_self.getAttributes(), (attribute, attributeName) => {
      if (!attribute.isSystemAttribute) {
        coreAttributeNames.push(attributeName)

        if (attribute.required && !attribute.defaultValue) {
          requiredAttributeNames.push(attributeName)
        }
      }
    })


    if (!this.mutations) {
      this.mutations = []
    }

    const mutationNames = []

    this.mutations.map((mutation) => {

      passOrThrow(
        !mutationNames.includes(mutation.name),
        () => `Duplicate mutation name '${mutation.name}' found in '${this.name}'`
      )

      mutationNames.push(mutation.name)

      if (mutation.attributes) {
        mutation.attributes.map((attributeName) => {
          passOrThrow(
            this._attributes[ attributeName ],
            () => `Cannot use attribute '${this.name}.${attributeName}' in mutation '${this.name}.${mutation.name}' as it does not exist`
          )
        })

        if (mutation.type === MUTATION_TYPE_CREATE) {
          const missingAttributeNames = requiredAttributeNames.filter(requiredAttributeName => {
            return !mutation.attributes.includes(requiredAttributeName)
          })

          passOrThrow(
            missingAttributeNames.length === 0,
            () => `Missing required attributes in mutation '${this.name}.${mutation.name}' need to have a defaultValue() function: [ ${missingAttributeNames.join(', ')} ]`
          )
        }
      }
    })

    defaultEntityMutations.map(defaultMutation => {
      if (!mutationNames.includes(defaultMutation.name)) {
        this.mutations.push(new Mutation({
          name: defaultMutation.name,
          type: defaultMutation.type,
          description: defaultMutation.description(this.name),
          attributes: coreAttributeNames
        }))
      }
    })

  }


  referencedBy (sourceEntityName, sourceAttributeName) {
    passOrThrow(
      sourceEntityName,
      () => `Entity '${this.name}' expects an entity to be referenced by`
    )

    passOrThrow(
      sourceAttributeName,
      () => `Entity '${this.name}' expects a source attribute to be referenced by`
    )

    let found = false;

    this.referencedByEntities.map(entry => {
      if (entry.sourceEntityName === sourceEntityName  &&  entry.sourceAttributeName === sourceAttributeName ) {
        found = true
      }
    })

    if (!found) {
      this.referencedByEntities.push({
        sourceEntityName,
        sourceAttributeName,
      })
    }
  }


  toString() {
    return this.name
  }

}


export default Entity


export const isEntity = (obj) => {
  return (obj instanceof Entity)
}
