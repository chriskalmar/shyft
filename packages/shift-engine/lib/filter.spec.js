
import {
  splitAttributeAndFilterOperator,
  processFilterLevel
} from './filter'

import Entity from './entity/Entity';

import {
  DataTypeBoolean,
  DataTypeString,
  DataTypeInteger,
} from './datatype/dataTypes';

import StorageType from './storage/StorageType';
import StorageDataType from './storage/StorageDataType';



describe('filter', () => {

  const filteredEntity = new Entity({
    name: 'FilteredEntityName',
    description: 'Just some description',
    attributes: {
      id: {
        type: DataTypeInteger,
        description: 'ID of user',
        isPrimary: true,
      },
      firstName: {
        type: DataTypeString,
        description: 'First name'
      },
      lastName: {
        type: DataTypeString,
        description: 'Last name'
      },
      isActive: {
        type: DataTypeBoolean,
        description: 'User has been active within this month'
      },
    }
  })


  const SomeStorageType = new StorageType({
    name: 'SomeStorageType',
    description: 'Just some description',
    findOne() { },
    findOneByValues() { },
    find() { },
    count() { },
    mutate() { },
  })

  const StorageDataTypeAny = new StorageDataType({
    name: 'StorageDataTypeAny',
    description: 'Just some description',
    nativeDataType: 'text',
    serialize() { },
    capabilities: [
      'lt',
      'lte',
      'gt',
      'gte',
    ]
  })

  const StorageDataTypeText = new StorageDataType({
    name: 'StorageDataTypeText',
    description: 'Just some description',
    nativeDataType: 'text',
    serialize() { },
    capabilities: [
      'in',
      'lt',
      'lte',
      'gt',
      'gte',
      'starts_with',
      'ends_with',
    ],
  })


  SomeStorageType.addDataTypeMap(DataTypeInteger, StorageDataTypeAny)
  SomeStorageType.addDataTypeMap(DataTypeBoolean, StorageDataTypeAny)
  SomeStorageType.addDataTypeMap(DataTypeString, StorageDataTypeText)



  describe('splitAttributeAndFilterOperator', () => {


    it('should split attributes from operators', () => {

      expect(splitAttributeAndFilterOperator('test')).toEqual({ attributeName: 'test' })
      expect(splitAttributeAndFilterOperator('__test')).toEqual({ attributeName: '__test' })
      expect(splitAttributeAndFilterOperator('login__lt')).toEqual({ attributeName: 'login', operator: 'lt' })
      expect(splitAttributeAndFilterOperator('firstName__gte')).toEqual({ attributeName: 'firstName', operator: 'gte' })
      expect(splitAttributeAndFilterOperator('last_name__ne')).toEqual({ attributeName: 'last_name', operator: 'ne' })
      expect(splitAttributeAndFilterOperator('some__long_attribute__name__lte')).toEqual({ attributeName: 'some__long_attribute__name', operator: 'lte' })
      expect(splitAttributeAndFilterOperator('__some_name__ne')).toEqual({ attributeName: '__some_name', operator: 'ne' })
      expect(splitAttributeAndFilterOperator('___some_name__ne')).toEqual({ attributeName: '___some_name', operator: 'ne' })
      expect(splitAttributeAndFilterOperator('___some_name___ne')).toEqual({ attributeName: '___some_name_', operator: 'ne' })

    })


    it('should fail on wrong inputs', () => {


      function fn1() {
        splitAttributeAndFilterOperator()
      }

      function fn2() {
        splitAttributeAndFilterOperator([])
      }

      function fn3() {
        splitAttributeAndFilterOperator({ any: 'thing' })
      }

      function fn4() {
        splitAttributeAndFilterOperator('name__')
      }

      function fn5() {
        splitAttributeAndFilterOperator('name___')
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
      expect(fn4).toThrowErrorMatchingSnapshot();
      expect(fn5).toThrowErrorMatchingSnapshot();

    })

  })



  describe('processFilterLevel', () => {


    it('should process filter level', () => {

      const goodFilter1 = {
        lastName: 'Doe',
        firstName__gte: 'J',
      }

      const result1 = {
        lastName: 'Doe',
        firstName: {
          $gte: 'J'
        }
      }


      const goodFilter2 = {
        lastName__in: ['Doe', 'Smith'],
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        isActive: true,
      }

      const result2 = {
        lastName: {
          $in: ['Doe', 'Smith'],
        },
        firstName: {
          $starts_with: 'Joh',
          $ends_with: 'an',
        },
        isActive: true,
      }


      expect(
        processFilterLevel(goodFilter1, filteredEntity.getAttributes(), ['somewhere'], SomeStorageType)
      ).toEqual(result1)

      expect(
        processFilterLevel(goodFilter2, filteredEntity.getAttributes(), ['somewhere'], SomeStorageType)
      ).toEqual(result2)

    })


    it('should process nested filter levels', () => {

      const goodFilter1 = {
        lastName: 'Doe',
        OR: [
          {
            firstName: 'Jack'
          },
          {
            firstName__starts_with: 'A',
            isActive: true
          }
        ]
      }

      const result1 = {
        lastName: 'Doe',
        $or: [
          {
            firstName: 'Jack'
          },
          {
            firstName: {
              $starts_with: 'A'
            },
            isActive: true
          }
        ]
      }


      const goodFilter2 = {
        lastName__gt: 'Tomson',
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        AND: [
          {
            lastName__starts_with: 'Und',
            lastName__ends_with: 'ton',
          }
        ],
        isActive: true,
      }

      const result2 = {
        lastName: {
          $gt: 'Tomson'
        },
        firstName: {
          $starts_with: 'Joh',
          $ends_with: 'an',
        },
        $and: [
          {
            lastName: {
              $starts_with: 'Und',
              $ends_with: 'ton',
            }
          }
        ],
        isActive: true,
      }


      expect(
        processFilterLevel(goodFilter1, filteredEntity.getAttributes(), ['somewhere'], SomeStorageType)
      ).toEqual(result1)

      expect(
        processFilterLevel(goodFilter2, filteredEntity.getAttributes(), ['somewhere'], SomeStorageType)
      ).toEqual(result2)

    })


    it('should throw if provided params are invalid', () => {

      function fn1() {
        processFilterLevel()
      }

      function fn2() {
        processFilterLevel([], null, [])
      }

      function fn3() {
        processFilterLevel([], null, ['somewhere'], SomeStorageType)
      }

      function fn4() {
        processFilterLevel([], null, ['somewhere', 'deeply', 'nested'], SomeStorageType)
      }

      function fn5() {
        processFilterLevel({}, null, ['somewhere', 'deeply', 'nested'], SomeStorageType)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
      expect(fn4).toThrowErrorMatchingSnapshot();
      expect(fn5).toThrowErrorMatchingSnapshot();

    })


    it('should throw if invalid attributes are used in filter', () => {

      const badFilter1 = {
        lastName: 'Doe',
        firstName__gte: 'J',
        something__ne: true,
      }

      const badFilter2 = {
        lastName: 'Doe',
        firstName__gte: 'J',
        anything_here: 'test',
      }


      function fn1() {
        processFilterLevel(badFilter1, filteredEntity.getAttributes(), null, SomeStorageType)
      }

      function fn2() {
        processFilterLevel(badFilter2, filteredEntity.getAttributes(), null, SomeStorageType)
      }

      function fn3() {
        processFilterLevel(badFilter2, filteredEntity.getAttributes(), ['just', 'here'], SomeStorageType)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    })


    it('should throw if invalid operators are used in filter', () => {

      const badFilter1 = {
        lastName: 'Doe',
        isActive__ends_with: 'J',
      }

      const badFilter2 = {
        lastName: 'Doe',
        firstName__anything: 'J',
      }


      function fn1() {
        processFilterLevel(badFilter1, filteredEntity.getAttributes(), null, SomeStorageType)
      }

      function fn2() {
        processFilterLevel(badFilter2, filteredEntity.getAttributes(), null, SomeStorageType)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
    })


    it('should throw if exact match operators is used with another operator on the same attribute', () => {

      const badFilter1 = {
        lastName__in: ['Doe', 'Smith'],
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        firstName: 'Frank',
        isActive: true,
      }

      const badFilter2 = {
        lastName__in: ['Doe', 'Smith'],
        firstName: 'Frank',
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        isActive: true,
      }

      function fn1() {
        processFilterLevel(badFilter1, filteredEntity.getAttributes(), null, SomeStorageType)
      }

      function fn2() {
        processFilterLevel(badFilter2, filteredEntity.getAttributes(), null, SomeStorageType)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();

    })

  })

})
