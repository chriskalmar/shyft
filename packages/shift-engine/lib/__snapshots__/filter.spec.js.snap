// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter processFilterLevel should throw if exact match operators is used with another operator on the same attribute 1`] = `"Cannot combine 'exact match' operator with other operators on attribute 'firstName' used in filter"`;

exports[`filter processFilterLevel should throw if exact match operators is used with another operator on the same attribute 2`] = `"Cannot combine 'exact match' operator with other operators on attribute 'firstName' used in filter"`;

exports[`filter processFilterLevel should throw if invalid attributes are used in filter 1`] = `"Unknown attribute name 'something' used in filter"`;

exports[`filter processFilterLevel should throw if invalid attributes are used in filter 2`] = `"Unknown attribute name 'anything_here' used in filter"`;

exports[`filter processFilterLevel should throw if invalid attributes are used in filter 3`] = `"Unknown attribute name 'anything_here' used in filter at 'just.here'"`;

exports[`filter processFilterLevel should throw if invalid operators are used in filter 1`] = `"Unknown or incompatible operator 'ends_with' used on 'isActive' in filter"`;

exports[`filter processFilterLevel should throw if invalid operators are used in filter 2`] = `"Unknown or incompatible operator 'anything' used on 'firstName' in filter"`;

exports[`filter processFilterLevel should throw if provided params are invalid 1`] = `"filter needs to be an object of filter criteria"`;

exports[`filter processFilterLevel should throw if provided params are invalid 2`] = `"optional path in processFilterLevel() needs to be an array"`;

exports[`filter processFilterLevel should throw if provided params are invalid 3`] = `"filter at 'somewhere' needs to be an object of filter criteria"`;

exports[`filter processFilterLevel should throw if provided params are invalid 4`] = `"filter at 'somewhere.deeply.nested' needs to be an object of filter criteria"`;

exports[`filter processFilterLevel should throw if provided params are invalid 5`] = `"processFilterLevel() expects an attribute map"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 1`] = `"invalid filter 'undefined'"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 2`] = `"invalid filter ''"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 3`] = `"invalid filter '[object Object]'"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 4`] = `"invalid filter 'name__'"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 5`] = `"invalid filter 'name___'"`;
